### PR TITLE
Fix GCP VM leak issue

### PR DESCRIPTION
**Pull request by infwinston**
Fix a long existing resource leak issue on GCP triggered by `test_stale_job` smoke test.

This bug happened because GCP doesn't fully stop an instance even after a stop operation is finished.

This PR make `sky stop` wait until the instance actually stopped.
I tried to benchmark whether adding this status check for loop increases the running time of `sky stop`.
Turns out the running time is almost the same as before (~40s) for normal use cases. Only if you create a VM and immediately stop it then it'd take 2mins to finish (like what happened in our smoke test).

On the other hand, I also add safe guard to `sky start` to avoid launching new VM. Now if user `sky start` a STOPPING VM (this only happens when user stops the VM on web console) then it'd return below error message.
```
sky start test-stale-job-4da3f903-f7 -y
...
I 08-18 22:28:32 cloud_vm_ray_backend.py:1075] Launching on GCP us-central1 (us-central1-a)
D 08-18 22:29:34 cloud_vm_ray_backend.py:1137] Ray up takes 62.20364475250244 seconds with 1 retries.
W 08-18 22:29:34 cloud_vm_ray_backend.py:485] Got return code RESOURCE_NOT_READY in us-central1-a (message: The resource 'projects/intercloud-320520/zones/us-central1-a/instances/ray-test-s
tale-job-4da3f903-f7-head-4819c50c-compute' is not ready)
sky.exceptions.ResourcesUnavailableError: Failed to acquire resources to restart the stopped cluster test-stale-job-4da3f903-f7 on Region(name='us-central1'). Please retry again later.
To keep retrying until the cluster is up, use the `--retry-until-up` flag.
```

- [x] passed `test_stale_job` smoke test

### CODE REVIEW

File: sky/backends/cloud_vm_ray_backend.py [hunk 1]
```diff
@@ -508,6 +508,9 @@ def _update_blocklist_on_gcp_error(self, region, zones, stdout, stderr):
                     # However, UNSUPPORTED_OPERATION is observed empirically when VM is preempted during creation.
                     # This seems to be not documented by GCP.
                     self._blocked_zones.add(zone.name)
+                elif code in ['RESOURCE_NOT_READY']:
+                    # This code is returned when the VM is still STOPPING.
+                    self._blocked_zones.add(zone.name)
```
Comment (concretevitamin) [inline sky/backends/cloud_vm_ray_backend.py:513]: Why do we block the zone in this case?  Is it possible that
- a VM is STOPPING
- a concurrent `sky start` somehow hits this branch

From users' perspectives, the `sky start` should start the same VM in its original zone?

File: sky/skylet/providers/gcp/node_provider.py [hunk 2]
```diff
@@ -191,11 +193,11 @@ def create_node(self, base_config: dict, tags: dict, count: int) -> None:
                     filters[TAG_RAY_USER_NODE_TYPE] = labels[TAG_RAY_USER_NODE_TYPE]
                 # SKY: "TERMINATED" for compute VM, "STOPPED" for TPU VM
```
Comment (concretevitamin) [inline sky/skylet/providers/gcp/node_provider.py:194]: This comment can be updated.

File: sky/skylet/providers/gcp/node_provider.py [hunk 3]
```diff
@@ -191,11 +193,11 @@ def create_node(self, base_config: dict, tags: dict, count: int) -> None:
...
                 if isinstance(resource, GCPCompute):
-                    STOPPED_STATUS = ["TERMINATED"]
+                    TO_BE_STARTED_STATUS = ["TERMINATED", "STOPPING"]
                 else:
-                    STOPPED_STATUS = ["STOPPED"]
+                    TO_BE_STARTED_STATUS = ["STOPPED", "STOPPING"]
                 reuse_nodes = resource._list_instances(
-                    filters, STOPPED_STATUS)[:count]
+                    filters, TO_BE_STARTED_STATUS)[:count]
```
Comment (concretevitamin) [inline sky/skylet/providers/gcp/node_provider.py:200]: The new name suggests this is going to be started, but not necessarily.  I think the old name STOPPED_STATUS is more accurate.

File: sky/skylet/providers/gcp/node_provider.py [hunk 4]
```diff
@@ -229,6 +231,21 @@ def terminate_node(self, node_id: str):
...
+                        instance = resource.get_instance(node_id=node_id)
+                        if instance.is_stopped():
+                            logger.info(f'Instance {node_id} is stopped.')
+                            break
+                        elif instance.is_stopping():
+                            time.sleep(POLL_INTERVAL)
+                        else:
+                            raise RuntimeError(f'Unexpected instance status. Details: {instance}')
+
+                    if instance.is_stopping():
```
Comment (concretevitamin) [inline sky/skylet/providers/gcp/node_provider.py:247]: ...is still in STOPPING status.

**Review by concretevitamin — COMMENTED** [review]
This is awesome @infwinston! Can we also run all smoke tests - just to cover all tests using GCP?

**Comment by concretevitamin** [comment]
> On the other hand, I also add safe guard to `sky start` to avoid launching new VM. Now if user `sky start` a STOPPING VM (this only happens when user stops the VM on web console) then it'd return below error message.

Which module is this safeguard added to?

File: sky/backends/cloud_vm_ray_backend.py (hunk 1 above)
Comment (infwinston) [reply sky/backends/cloud_vm_ray_backend.py:513]: As explained offline, failover does not apply here. so `sky start` would just fail if this error code is returned.

File: sky/skylet/providers/gcp/node_provider.py (hunk 2 above)
Comment (infwinston) [reply sky/skylet/providers/gcp/node_provider.py:194]: Updated!

File: sky/skylet/providers/gcp/node_provider.py (hunk 3 above)
Comment (infwinston) [reply sky/skylet/providers/gcp/node_provider.py]: Fixed!

File: sky/skylet/providers/gcp/node_provider.py (hunk 4 above)
Comment (infwinston) [reply sky/skylet/providers/gcp/node_provider.py:249]: Sorry I don't quite get this comment?

**Comment by infwinston** [comment]
> Which module is this safeguard added to?

As explained offline, as `STOPPING` status is included, `sky start` would not try to launch a new VM.

**Comment by infwinston** [comment]
Smoke tests all passed except for `test_spot_storage` which seems to be unrelated to this PR I assume?

File: sky/skylet/providers/gcp/node_provider.py (hunk 4 above)
Comment (concretevitamin) [reply sky/skylet/providers/gcp/node_provider.py:249]: Update error message on 248: "is still not terminated" -> "is still in STOPPING status"

**Review by concretevitamin — APPROVED** [review]
LGTM

File: sky/skylet/providers/gcp/node_provider.py (hunk 4 above)
Comment (infwinston) [reply sky/skylet/providers/gcp/node_provider.py:251]: Got it.

**Pull request merged**